week04/CopyListWithRandomPointer.java

```diff
@@ -0,0 +1,41 @@
+/*
+// Definition for a Node.
+class Node {
+    int val;
+    Node next;
+    Node random;
+
+    public Node(int val) {
+        this.val = val;
+        this.next = null;
+        this.random = null;
+    }
+}
+*/
+
+// TC = O(N); where N is the number of nodes;
+// MC = O(N);
+// Problem link : https://leetcode.com/problems/copy-list-with-random-pointer/description/
+
+class Solution {
+
+    public Node copyRandomList(Node head) {
+        Map<Node, Node>copies = new HashMap<>();
```

Comment:
Map does not support constant access. So the overall time complexity would be `O(NlogN)`.

Reply:
Vai, Here Map is the interface whose implementation is the HashMap. Then will it be O(NlogN)?